src/analytics/private/segments/user-segments-workflow.ts

```diff
@@ -0,0 +1,202 @@
+import { RetentionDays } from 'aws-cdk-lib/aws-logs';
+import { Bucket } from 'aws-cdk-lib/aws-s3';
+import {
+  Choice,
+  Condition,
+  Fail,
+  IStateMachine,
+  LogLevel,
+  StateMachine,
+  Succeed,
+  TaskInput,
+  Wait,
+  WaitTime,
+} from 'aws-cdk-lib/aws-stepfunctions';
+import { LambdaInvoke } from 'aws-cdk-lib/aws-stepfunctions-tasks';
+import { Construct } from 'constructs';
+import { SegmentJobStatus } from './segments-model';
+import { createLambdaRole } from '../../../common/lambda';
+import { createLogGroup } from '../../../common/logs';
+import { REDSHIFT_MODE } from '../../../common/model';
+import { SolutionNodejsFunction } from '../../../private/function';
+import { ExistingRedshiftServerlessProps, ProvisionedRedshiftProps } from '../model';
+
+export interface UserSegmentsWorkflowProps {
+  readonly projectId: string;
+  readonly networkConfig: {
+    readonly vpc: IVpc;
+    readonly vpcSubnets: SubnetSelection;
+  };
+  readonly securityGroupForLambda: ISecurityGroup;
+  readonly clickstreamMetadataDdbTable: ITable;
+  readonly dataAPIRole: IRole;
+  readonly serverlessRedshift?: ExistingRedshiftServerlessProps;
+  readonly provisionedRedshift?: ProvisionedRedshiftProps;
+  readonly databaseName: string;
+  readonly pipelineS3Bucket: string;
+  readonly segmentsS3Prefix: string;
+}
+
+const lambdaRootPath = __dirname + '/../../lambdas/user-segments-workflow';
+
+export class UserSegmentsWorkflow extends Construct {
+  private readonly props: UserSegmentsWorkflowProps;
+  public readonly userSegmentsWorkflow: IStateMachine;
+
+  constructor(scope: Construct, id: string, props: UserSegmentsWorkflowProps) {
+    super(scope, id);
+
+    this.props = props;
+    this.userSegmentsWorkflow = this.createWorkflow(props);
+  }
+
+  private createWorkflow(props: UserSegmentsWorkflowProps): IStateMachine {
+    // Define task for segment job initialization
+    const segmentJobInitFunc = this.constructNodejsFunction('segment-job-init', [
+      new PolicyStatement({
+        actions: ['events:DisableRule'],
+        resources: [`arn:${Aws.PARTITION}:events:*:${Aws.ACCOUNT_ID}:rule/Clickstream-*`],
+      }),
+    ], {
+      CLICKSTREAM_METADATA_DDB_ARN: props.clickstreamMetadataDdbTable.tableArn,
+    });
+    const segmentJobInitTask = new LambdaInvoke(this, 'WorkflowTask-SegmentJobInit', {
+      lambdaFunction: segmentJobInitFunc,
+      outputPath: '$.Payload',
+    });
+    props.clickstreamMetadataDdbTable.grantReadWriteData(segmentJobInitFunc);
+
+    // Define task for checking state machine status
+    const stateMachineStatusTask = new LambdaInvoke(this, 'WorkflowTask-StateMachineStatus', {
+      lambdaFunction: this.constructNodejsFunction('state-machine-status', [
+        new PolicyStatement({
+          actions: ['states:ListExecutions'],
+          resources: [`arn:${Aws.PARTITION}:states:*:${Aws.ACCOUNT_ID}:stateMachine:ClickstreamUserSegmentsWorkflowStateMachine*`],
```

Comment:
You might try this.

https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_stepfunctions-readme.html#execution-level-permissions

Reply:
done

Reply:
done

Reply:
Could you create a separate role for this Lamba, then grant the role access to the executions of the state machine?